src/Hedgehog.Experimental/Gen.fs

```diff
@@ -525,33 +548,42 @@ module GenX =
               return values.GetValue index |> unbox
             }
 
+        | Shape.Collection s ->
+            s.Accept {
+              new ICollectionVisitor<Gen<'a>> with
+              member _.Visit<'collection, 'element when 'collection :> System.Collections.Generic.ICollection<'element>> () =
+                match typeShape with
+                | Shape.Poco (:? ShapePoco<'a> as shape) ->
+                  gen {
+                    let! length =
+                      if canRecurse typeof<'element> then
+                        Gen.integral config.SeqRange
+                      else
+                        0 |> Gen.constant
+                    let! elements =
+                      incrementRecursionDepth typeof<'element>
+                      |> autoInner config
+                      |> List.replicate length
+                      |> ListGen.sequence
+                    let! collection =
+                      genPoco shape
+                      |> Gen.map unbox<System.Collections.Generic.ICollection<'element>>
```

Comment:
This passes my test, but I don't that the constructor parameters should be generated last.  We just need the constructor call to be last.

We need to generate three things:
- constructor parameters,
- `length` (which I want to rename to `count`), and
- the elements.

The elements depend on the length but the constructor parameters are independent of both the length and the elements, so they should be "combined applicatively" (once that capability exists).  In the meantime, I think it makes more sense to obtain the parameters first.

This change doesn't affect my test because `ResizeArray` has a default (i.e. 0-argument) constructor, so the constructor parameter generation is a vacuous step.

Reply:
I'm not seeing how I can (easily) pull the generation of ctorArgs to the level of this gen. I'm assuming here that `genPoco` stays more or less the same. Am I misunderstanding?

Reply:
Change `genPoco` to have return type
```fsharp
Gen<'ctorParams> * ('ctorParam -> 'a)
```

Reply:
Hm, as your above code reveals, this causes the `'ctorParams` to "escape" the scope of `genPoco`. The caller doesn't know the type of `'ctorParams` - we only have it because `IConstructorVisitor` gives it to us, and that's only visible from the scope of `genPoco`. We could just box it up... but I'm hesitant to lose type safety and introduce complexity.

Reply:
Your `unit` suggestion works.